src/external/ACSAdapter/AMSFileScanner.ts

```diff
@@ -0,0 +1,138 @@
+import FileMetadata from "@microsoft/omnichannel-amsclient/lib/FileMetadata";
+import FramedClient from "@microsoft/omnichannel-amsclient/lib/FramedClient";
+import sleep from "../../utils/sleep";
+import { AMSDownloadStatus, defaultScanInterval } from "./AMSFileManager";
+import activityUtils from "./activityUtils";
+
+interface FileScanResponse {
+    status: string;
+}
+
+interface FileScanResult {
+    fileMetadata: FileMetadata;
+    scan: FileScanResponse;
+    next?: any; // eslint-disable-line @typescript-eslint/no-explicit-any
+    activity?: any; // eslint-disable-line @typescript-eslint/no-explicit-any
+}
+
+class AMSFileScanner {
+    public amsClient: FramedClient;
+    public scanResults: Map<string, FileScanResult> | null = null;
+    public interval: number = defaultScanInterval;
+    private shouldQueueScan: boolean;
+
+    constructor(amsClient: FramedClient, interval: number = defaultScanInterval) {
+        this.amsClient = amsClient;
+        this.scanResults = new Map<string, FileScanResult>();
+        this.interval = interval;
+        this.shouldQueueScan = false;
+        this.queueScan();
+    }
+
+    public async queueScan(): Promise<void> {
+        this.shouldQueueScan = true;
+
+        try {
+            await this.scanFiles();
+        } catch (e) {
+            console.error(e);
+        } finally {
+            await sleep(this.interval);
+            this.shouldQueueScan && await this.queueScan();
+        }
+    }
+
+    public retrieveFileScanResult(id: string): FileScanResult | undefined {
+        return this.scanResults?.get(id);
+    }
+
+    public addOrUpdateFile(id: string, fileMetadata: FileMetadata, scan: FileScanResponse): void {
+        const scanResult = this.retrieveFileScanResult(id);
+        this.scanResults?.set(id, {
+            ...(scanResult || {}),
+            fileMetadata,
+            scan
+        });
+    }
+
+    public addNext(id: string, next: Function): void {
+        const scanResult = this.scanResults?.get(id);
+        if (scanResult) {
+            this.scanResults?.set(id, {...scanResult, next});
+        }
+    }
+
+    public addActivity(id: string, activity: any): void {  // eslint-disable-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
+        const scanResult = this.scanResults?.get(id);
+        if (scanResult) {
+            this.scanResults?.set(id, {...scanResult, activity});
+        }
+    }
+
+    public async scanFileCallback(scanResult: FileScanResult, id: string): Promise<void> {

```

Comment:
IMO: this method is too large, my suggestion is to break it into functional parts, in this way, it makes the code easier to debug and spot any possible failure,